[RNMobile] - E2E Simplify heading and lists blocks functions

jostnes opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
- This PR is a follow-up to https://github.com/WordPress/gutenberg/pull/40404, this PR will attempt to simplify other text block implementations. I earlier missed that it was implemented separately for paragraph/heading/list and did not want to introduce more changes to the previous PR.
- With the changes, we can also now re-enable the slash inserter tests and close https://github.com/wordpress-mobile/gutenberg-mobile/issues/3699
- Also included in the PR is rearranging the location of some functions in the `editor-page.js` so that it would be easier to find

## Why?
To reduce E2E test flakiness when running in the CI. Also to simplify the implementation, we can now use 1 function instead of many for text-based blocks. Also will be able to remove the extra click that was previously needed on Android on the test level, as it has since been covered on the function itself.

## How?
For Paragraph and Heading blocks, we can now use `getTextBlockAtPosition()` to get block position. Would still have to separate the List block because of locator differences. To access List block, use `getListBlock()`. To type text into blocks, all text blocks can now use `typeTextToTextBlock()`

## Testing Instructions
Test impacted in this PR should now work both locally and in the CI on both platforms (change platform from `android` to `ios` to test iOS)
```
TEST_RN_PLATFORM=android npm run native device-tests gutenberg-editor-heading-@canary.test
TEST_RN_PLATFORM=android npm run native device-tests gutenberg-editor-lists-@canary.test
TEST_RN_PLATFORM=android npm run native device-tests gutenberg-editor-lists-end.test.js
TEST_RN_PLATFORM=android npm run native device-tests gutenberg-editor-lists.test.js
TEST_RN_PLATFORM=android npm run native device-tests gutenberg-editor-paragraph.test
TEST_RN_PLATFORM=android npm run native device-tests gutenberg-editor-paste.test
TEST_RN_PLATFORM=android npm run native device-tests gutenberg-editor-rotation.test
TEST_RN_PLATFORM=android npm run native device-tests gutenberg-editor-slash-inserter-@canary.test
```
## Screenshots or screencast <!-- if applicable -->

jostnes committed:
update tests using paragraph, heading and list blocks

jostnes committed:
fix slash inserter tests to work in ci

jostnes committed:
lint fixes

jostnes committed:
wait for ordered list to appear

jostnes committed:
lint fixes

jostnes committed:
extra click only on local env

jostnes committed:
wait to get backspace click reflected

jostnes committed:
re-add extra click only for local env

jostnes committed:
add wait to wait for backspace key to be reflected

jostnes committed:
lint fixes

jostnes commented on packages/react-native-editor/__device-tests__/helpers/utils.js:
```diff
@@ -480,7 +496,16 @@ const waitForVisible = async ( driver, elementLocator, iteration = 0 ) => {
+	if ( returnBool ) {
+		return true;
 	}
 	return locator[ 0 ];
```
Updated this to return two different types. Not sure if the best way to do this, should this be separated instead?

fluiddot replied:
Yep, I'd separate it into a new function, as it can be a bit confusing to have two different return types in this function.

jostnes commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -2,17 +2,17 @@
-	isEditorVisible,
+	swipeUp,
+	toggleHtmlMode,
+	typeString,
```
No actual change here, only sort order.

jostnes commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -64,6 +68,10 @@ class EditorPage {
 		return await waitForVisible( this.driver, blockLocator );
 	}
 
+	async typeTextToTextBlock( block, text, clear ) {
```
Moved this up to be grouped together under `Text blocks functions`

jostnes commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -542,35 +542,60 @@ class EditorPage {
 		return await blockLocator.text();
 	}
 
+	async getNumberOfParagraphBlocks() {
```
Moved this up to be grouped together under `Paragraph Block functions`.

jostnes commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -542,35 +542,60 @@ class EditorPage {
+		return locator.length;
+	}
+
+	async assertSlashInserterPresent() {
```
Moved this from `./gutenberg-editor-slash-inserter-@canary.test.js` to keep the test file cleaner.

jostnes commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -760,34 +757,25 @@ class EditorPage {
 
 const blockNames = {
-	paragraph: 'Paragraph',
-	gallery: 'Gallery',
+	audio: 'Audio',
 	columns: 'Columns',
 	cover: 'Cover',
+	embed: 'Embed',
+	file: 'File',
+	gallery: 'Gallery',
 	heading: 'Heading',
 	image: 'Image',
 	latestPosts: 'Latest Posts',
 	list: 'List',
 	more: 'More',
+	paragraph: 'Paragraph',
+	search: 'Search',
 	separator: 'Separator',
 	spacer: 'Spacer',
 	verse: 'Verse',
```
No actual change here, only sort order.

fluiddot commented on packages/react-native-editor/__device-tests__/helpers/utils.js:
```diff
@@ -464,11 +464,27 @@ const waitForMediaLibrary = async ( driver ) => {
+ *
+ * @param {string} driver
+ * @param {string} elementLocator
+ * @param {boolean} returnBool - To return boolean instead of the first element/throw error. Useful for asserts.
```
I totally agree that it would be helpful to have a helper function to check if an element is visible, however, I'm wondering if we should expose that functionality within this function or create a new one. In general, I rather have functions that provide a single functionality instead of multiple as it can make it harder to understand how it works. Being this said, what do you think about creating a new helper like `isElementVisible` to expose the functionality?

jostnes replied:
Yeah, I was conflicted about this one since I only needed the bool return value in one case, but I do agree to have it do multiple things can make it harder to understand. Thanks for your input! I will separate it and create a new `isElementVisible` that will return bool.

jostnes replied:
This is done in this commit: https://github.com/WordPress/gutenberg/pull/40670/commits/46273a83845071272a8950c58365a30608581a1e

fluiddot commented on packages/react-native-editor/__device-tests__/helpers/utils.js:
```diff
@@ -464,11 +464,27 @@ const waitForMediaLibrary = async ( driver ) => {
+	returnBool,
+	iteration = 0
+) => {
+	const maxIteration = returnBool ? 5 : 25;
```
I'm curious about why using `returnBool` argument implies having a lower maximum iteration, do those calls require less iterations to determine if the element is visible?

jostnes replied:
Ah right, sorry it's not clear, in the one case that relied on the bool value, it was waiting for the element to not be visible instead of waiting for it to be visible. 5 was a magic number since I thought it didn't need to wait that long to validate that. 

It probably doesn't need to wait 25 seconds for that, I think 5 seconds is enough, what do you think?

fluiddot replied:
Oh ok, that makes sense. In that case, we might consider setting the max iteration value via an argument, this could be helpful for example if `isElementVisible` function uses this helper and wants to use a lower max iteration to reduce the waiting.

jostnes replied:
That's a good point, it would make it more flexible. I've updated it in this commit: https://github.com/WordPress/gutenberg/pull/40670/commits/85c407c286fc18e226f98fbcfa440237f995d093

fluiddot commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -542,35 +542,60 @@ class EditorPage {
+				this.driver,
+				listBlockLocator
+			);
+			await textBlock.click();
```
Nitpick: since we're querying a list block, I'd expect the name if this constant to be `listBlock` instead of `textBlock`.

fluiddot commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -542,35 +542,60 @@ class EditorPage {
-		const clear = false;
+		if ( isAndroid() ) {
+			listBlockTextLocator = `//android.view.ViewGroup[contains(@content-desc, "List Block.")]/android.widget.EditText`;
+		}
```
I'm wondering if we could rearrange this part in a different way to handle both platform cases:
```
const listBlockTextLocatorIOS = options.isEmptyBlock
	? `(//XCUIElementTypeStaticText[contains(@name, "List")])`
	: `//XCUIElementTypeButton[contains(@name, "List")]//XCUIElementTypeTextView`;
const listBlockTextLocator = isAndroid()
	? `//android.view.ViewGroup[contains(@content-desc, "List Block.")]/android.widget.EditText`
	: listBlockTextLocatorIOS;
```

jostnes replied:
Yup your suggestion is cleaner, I've updated it in this commit: https://github.com/WordPress/gutenberg/pull/40670/commits/46273a83845071272a8950c58365a30608581a1e

fluiddot commented on packages/react-native-editor/__device-tests__/gutenberg-editor-heading-@canary.test.js:
```diff
@@ -59,9 +51,12 @@ describe( 'Gutenberg Editor tests', () => {
 		);
+
+		// Assert that even though there are 5 blocks, there should only be 3 paragraph blocks
+		expect( await editorPage.getNumberOfParagraphBlocks() ).toEqual( 3 );
```
I'd like to suggest relying on Jest snapshots for the expectations of tests, this is something we could apply on every test actually. We could have something like this:
```suggestion
		expect( await editorPage.getHtmlContent() ).toMatchSnapshot();
```

This would ensure that we produced the right HTML output after the test. Here is an example of the snapshot file:
```js
exports[`Gutenberg Editor tests should be able to create a post with heading and paragraph blocks 1`] = `
"<!-- wp:heading -->
<h2>Lorem Ipsum</h2>
<!-- /wp:heading -->

<!-- wp:paragraph -->
<p>The finer continuum interprets the polynomial rabbit. When can the geology runs? An astronomer runs. Should a communist consent?</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>The finer continuum interprets the polynomial rabbit. When can the geology runs? An astronomer runs. Should a communist consent?</p>
<!-- /wp:paragraph -->

<!-- wp:heading -->
<h2>Lorem Ipsum</h2>
<!-- /wp:heading -->

<!-- wp:paragraph -->
<p>The finer continuum interprets the polynomial rabbit. When can the geology runs? An astronomer runs. Should a communist consent?</p>
<!-- /wp:paragraph -->"
`;
```

fluiddot replied:
I'm thinking that we might consider covering this in a future PR. It would also help with reducing considerably [the test data fixture file](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/__device-tests__/helpers/test-data.js).

jostnes replied:
I'm open to trying that and looks like it's used quite a bit already in the integration tests. But yeah I wouldn't want to cover it in this PR and better to be in a separate PR, I think this would also make the assertions more consistent (there are a few E2E tests that does not check the HTML output). 

Though one thing I'm unsure of is if we can use Jest with Appium (not tested).

fluiddot replied:
> Though one thing I'm unsure of is if we can use Jest with Appium (not tested).

We are using Jest for running the tests ([reference](https://github.com/wordpress-mobile/gutenberg-mobile/blob/528d17a497818c839bee9ea7a79b2523b3e8c4d8/package.json#L78)) so there shouldn't be any problem. The snapshots are taken from the output HTML retrieved by the `getHtmlContent` helper which returns a string. Since the logic will be handled on the Jest side, I don't foresee any potential conflict due to using Appium.

fluiddot commented on packages/react-native-editor/__device-tests__/gutenberg-editor-lists-@canary.test.js:
```diff
@@ -8,27 +8,28 @@ import testData from './helpers/test-data';
-		await editorPage.sendTextToListBlock( listBlockElement, '\n' );
+		await editorPage.typeTextToTextBlock( listBlockElement, '\n', false );
+
+		listBlockElement = await editorPage.getListBlock();
```
If the block was already retrieved in line `21`, do we need need to query it again?

jostnes replied:
Retested this and looks like it is indeed not needed, I've removed it in [this commit](https://github.com/WordPress/gutenberg/pull/40670/commits/46273a83845071272a8950c58365a30608581a1e#diff-53684945abd1075e62d91961d64621a1f8a9c5a3189330c3d5a6b2159f0513cb)

fluiddot replied:
Great, thanks for checking it 🙇 .

fluiddot commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -542,35 +542,60 @@ class EditorPage {
-		if ( isAndroid() ) {
-			textViewElementName = 'android.widget.EditText';
-		}
+	async getListBlock( options = { isEmptyBlock: false } ) {
```
I noticed that this function doesn't provide an argument for querying a specific List block index, it would return always the first one, right?. Not sure if we have a case where we have multiple List blocks, but it would be great to have this functionality for future cases, wdyt?

jostnes replied:
Yup you're right, currently, it doesn't look for the position and just returns the first one. Let me see if I can add the position to it.

jostnes replied:
Added position as a parameter with the default value of 1 in [this commit](https://github.com/WordPress/gutenberg/pull/40670/commits/46273a83845071272a8950c58365a30608581a1e#diff-4749eb3277e22380531b254004b270c0d603cc201321292f13572abf7509a6f0).

fluiddot commented on packages/react-native-editor/__device-tests__/gutenberg-editor-lists.test.js:
```diff
@@ -2,28 +2,38 @@
+		await editorPage.typeTextToTextBlock( listBlockElement, '\n', false );
+
+		// Click List block on Android.
+		// Only needed when testing in local environment
+		if ( isAndroid() && isLocalEnvironment() ) {
+			await listBlockElement.click();
+		}
```
I'm curious about why this is only needed in the local environment 🤔 , shouldn't we expect the same behavior in both? It might actually be an issue diverging the testing logic between both platforms as it would make it hard for debugging errors.

jostnes replied:
Yeah this is a weird one (considering I'm using the exact same emulators on both envs 🙃  ), and that's a valid concern, we might have to think about more platforms if this condition is added (iOS vs Android vs Local vs CI). 

Let me try again if this can be removed so we don't complicate the condition.

fluiddot replied:
Great, thanks for checking it 🙇. Ideally, we should avoid having different logic between local and remote environments as both should behave equally. The reality is that we've experienced that each has its own nuances, so it might happen that there's no other way than to apply these types of tweaks.

fluiddot replied:
Resolving this conversion as I see that this comment was addressed in https://github.com/WordPress/gutenberg/pull/40670/commits/4f6917ea29be33315f89b96ea811278408390207.

fluiddot commented on packages/react-native-editor/__device-tests__/gutenberg-editor-lists.test.js:
```diff
@@ -2,28 +2,38 @@
+			false
+		);
+
+		// There is a delay in Sauce Labs when a key is sent
+		// There isn't an element to check as it's being typed into an element that already exists, workaround is to add this wait until there's a better solution
+		if ( isAndroid() && ! isLocalEnvironment() ) {
+			await editorPage.driver.sleep( 1500 );
+		}
```
I'm wondering if we should apply this workaround for other calls to `typeTextToTextBlock`, wdyt?

jostnes replied:
I think it would be better if it's here, as this is the one case I see this happening on, `typeTextToTextBlock` is used quite extensively so having a wait there might increase the test run time.

fluiddot replied:
Ok, I see. In that case, should we wait only when running the test in the remote environment? Should we apply this tweak in both?

jostnes replied:
Yup for this and the above lines 21-23, I've made some changes so that it applies to both environments, I created `waitIfAndroid();` for this with a wait time of 1 second for both environments, and looks to be working (latest CI is green and works on local too). This way we can remove the need to introduce `isLocalEnvironment` to this.

fluiddot replied:
Resolving this conversion as I see that this comment was addressed in https://github.com/WordPress/gutenberg/pull/40670/commits/4f6917ea29be33315f89b96ea811278408390207.

fluiddot reviewed (commented):
Thanks @jostnes for keeping improving the E2E tests 🏅  I added some comments within the code, let me know if you'd like me to expand on any of them.

Regarding tests, I ran them locally and found that on iOS, the test `gutenberg-editor-rotation.test.js` failed constantly with the following error:
```
[elementByAccessibilityId("Switch to HTML")] Error response status: 7, , NoSuchElement - An element could not be located on the page using the given search parameters. Selenium error: An element could not be located on the page using the given search parameters.
```

However, looks like it's related to probably running too fast (as I experienced on previous tests) as adding a wait in the test before getting the HTML makes it succeed 😅 .

jostnes committed:
break function, set position to get list block

jostnes committed:
lint fixes

jostnes committed:
use correct params, update function name

jostnes committed:
lint fixes

fluiddot commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -542,35 +543,70 @@ class EditorPage {
+		// Go to the correct list at position
+		const listBlockPositionLocator = isAndroid()
+			? `//android.view.ViewGroup[@content-desc="List Block. Row ${ position }"]`
+			: `(//XCUIElementTypeOther[contains(@name, "List Block. Row ${ position }")])[5]`;
```
Not sure if I read the Xpath properly, does the `[5]` part mean that it will look for the item at index `5`?

<img width="500" alt="Screenshot 2022-05-10 at 11 14 28" src="https://user-images.githubusercontent.com/14905380/167594127-684b8319-94e4-407a-aae9-db6719375932.png">

jostnes replied:
Whoops, my bad, I was testing with the number I got from Appium inspector and committed that 😬  I've updated this to 1 in the latest commit, the reason is that there multiple are returns and it needs to wait for at least 1 to be visible so 1 would be a safer choice rather than 5. 

Updated in this commit: https://github.com/WordPress/gutenberg/pull/40670/commits/d1af9076afc1a54787e003dd167aa1d17e0e1e21

fluiddot replied:
Great, thanks for addressing it 🙇 .

jostnes committed:
make maxIteration a parameter for isElementVisible

fluiddot commented on packages/react-native-editor/__device-tests__/helpers/utils.js:
```diff
@@ -482,30 +488,67 @@ const waitForVisible = async ( driver, elementLocator, iteration = 0 ) => {
+	}
+
+	return true;
+};
```
I noticed that the logic of this function is very similar to `waitForVisible`, so I'm wondering if we could call that function here instead of duplicating the logic, wdyt?

jostnes replied:
I'll try to see what I can do, in `waitForVisible()` an error is thrown if the element is not found causing the Appium session to be terminated (IIRC we couldn't use a try-catch because of this). Let me see if there's a workaround to that, but I do agree that the two are very similar except for the return values.

fluiddot replied:
Yep, most likely we could use `try-catch` here for checking if the element is visible.

jostnes replied:
I updated this so that `waitForVisible()` helper method does not throw an error when the element is not found, instead it will print an error and return an empty string (so the next step in the test will still fail if the element does not exist). I think this makes sense for 2 reasons:
1. Throwing an error in a test will immediately terminate the Appium session, so the test can't continue and if there are subsequent tests in the test suite, those tests will also fail
2. Noticed for a while now that if a test fails because it couldn't find the element, it's not reported properly in the tests tab in CircleCI ([example here](https://app.circleci.com/pipelines/github/wordpress-mobile/gutenberg-mobile/17320/workflows/886d74cc-4ddf-489b-a261-90d9447fa4fa/jobs/94374/steps))

The changes were done here: https://github.com/WordPress/gutenberg/pull/40670/commits/0077156bde9c865ea3b79d9bc398e13833531451

fluiddot replied:
> I updated this so that `waitForVisible()` helper method does not throw an error when the element is not found, instead it will print an error and return an empty string (so the next step in the test will still fail if the element does not exist). I think this makes sense for 2 reasons:
> 
> 1. Throwing an error in a test will immediately terminate the Appium session, so the test can't continue and if there are subsequent tests in the test suite, those tests will also fail
> 2. Noticed for a while now that if a test fails because it couldn't find the element, it's not reported properly in the tests tab in CircleCI ([example here](https://app.circleci.com/pipelines/github/wordpress-mobile/gutenberg-mobile/17320/workflows/886d74cc-4ddf-489b-a261-90d9447fa4fa/jobs/94374/steps))
> 
> The changes were done here: [0077156](https://github.com/WordPress/gutenberg/commit/0077156bde9c865ea3b79d9bc398e13833531451)

I like the idea of removing the exception, as you mentioned, this way we'll let the test case fail with the appropriate error if an element doesn't exist 👍 .

jostnes committed:
update xpath for list block

fluiddot commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -542,35 +543,70 @@ class EditorPage {
 
-	async sendTextToListBlock( block, text ) {
-		const textViewElement = await this.getTextViewForListBlock( block );
+		const listBlockTextLocatorIOS = options.isEmptyBlock
+			? `(//XCUIElementTypeStaticText[contains(@name, "List")])`
+			: `//XCUIElementTypeButton[contains(@name, "List")]//XCUIElementTypeTextView`;
```
This locator looks very similar to the one defined in lines 585 - 587, I'm wondering if we could reduce the code here by unifying both parts. The only difference I spotted is that when the block is not empty, here we're adding `//XCUIElementTypeTextView` at the end of the xpath, is this required? If so, then I understand that we need both locators.

https://github.com/WordPress/gutenberg/blob/85c407c286fc18e226f98fbcfa440237f995d093/packages/react-native-editor/__device-tests__/pages/editor-page.js#L585-L587

jostnes replied:
Ah yes, this is needed for list blocks, the locator for when the list is empty vs not empty is different. For both cases on iOS, we'll need to click on it first (lines 584 - 591). After clicking on it, when it's empty the XPath remains the same but when it's not empty, it will need `//XCUIElementTypeTextView` to get the correct element (lines 593 - 595).

fluiddot replied:
Ok, thanks for the clarification, in that case, let's continue having the two locators 👍 .

jostnes committed:
utilize waitForVisible for isElementVisible

jostnes committed:
lint fixes

jostnes commented:
Thank you for your consistently thorough review @fluiddot 🙇  

Just noticed your comment on a failing test:
> However, looks like it's related to probably running too fast (as I experienced on previous tests) as adding a wait in the test before getting the HTML makes it succeed 😅 .

I'll address that in the next PR where I'll be adding the wait to all blocks (and where it would make sense). After that is added hopefully we won't see flakiness caused by the test being too slow/fast anymore 🤞

fluiddot reviewed (approved):
LGTM 🎊 ! Thank you so much @jostnes for improving the E2E tests 🙇 .

I checked that all tests pass (including the full ones) in https://github.com/wordpress-mobile/gutenberg-mobile/pull/4805 ✅ .

jostnes committed:
Merge branch 'trunk' into rnmobile/heading-list-blocks-test-update

jostnes committed:
add wait to getNumberOfParagraphBlocks and update xpath for android list block

jostnes commented:
Hey @fluiddot, after merging the latest from trunk (to fix conflicts), some of the tests that were previously passing have started to fail consistently in the CI (locally tests are fine, which is a bit 🤯 ). 

It looked like a wait is needed on `getNumberOfParagraphBlocks()`, I added that and updated the XPath for the block list for Android because it was failing because the locator couldn't be found: `"//android.view.ViewGroup[contains(@content-desc, "List Block.")]/android.widget.EditText" is still not visible after 25 retries!` (good thing is, the tests are now being recorded as failed on the tests tab since we're no longer throwing an error when that happens 😄 )

I thought this commit would fix it: https://github.com/WordPress/gutenberg/pull/40670/commits/58d9e906c23fdbc26197a0ccaa9f3aad5a6b4147 but it did not...

I will keep this open for now, and request for review again once I figure out why this is happening/attempt to fix it.

fluiddot commented:
> Hey @fluiddot, after merging the latest from trunk (to fix conflicts), some of the tests that were previously passing have started to fail consistently in the CI (locally tests are fine, which is a bit 🤯 ).

I think the changes I recently introduced in https://github.com/WordPress/gutenberg/pull/40424 are most likely the culprit of the failures. In fact, I had to update some of the E2E tests in that PR because I also found some errors ([commit reference](https://github.com/WordPress/gutenberg/pull/40424/commits/decf119b85c3d1c89ef755936027476b909eae4a)).

The problem is related to the fact that the block's content is now wrapped over a view element, which breaks some of the queries that look for an `EditText` children element at the first level ([in this comment you can find further information related to this change](https://github.com/wordpress-mobile/gutenberg-mobile/pull/4832#issuecomment-1123616407)).

> It looked like a wait is needed on `getNumberOfParagraphBlocks()`, I added that and updated the XPath for the block list for Android because it was failing because the locator couldn't be found: `"//android.view.ViewGroup[contains(@content-desc, "List Block.")]/android.widget.EditText" is still not visible after 25 retries!` (good thing is, the tests are now being recorded as failed on the tests tab since we're no longer throwing an error when that happens 😄 )

If the issue is actually related to the queries, most likely it could be addressed by updating the Xpath so they look for the `EditText` element at any nested level. As an example of the locator you shared, I think we could do the following change:

```
//android.view.ViewGroup[contains(@content-desc, "List Block.")]/android.widget.EditText
```
=>
```
//android.view.ViewGroup[contains(@content-desc, "List Block.")]//android.widget.EditText
```

In relation to the changes in this PR, I think we should update the Android locators of the following lines:
- https://github.com/WordPress/gutenberg/blob/58d9e906c23fdbc26197a0ccaa9f3aad5a6b4147/packages/react-native-editor/__device-tests__/pages/editor-page.js#L548
- https://github.com/WordPress/gutenberg/blob/58d9e906c23fdbc26197a0ccaa9f3aad5a6b4147/packages/react-native-editor/__device-tests__/pages/editor-page.js#L597

@jostnes let me know if I can help out on addressing this, thanks 🙇 !

jostnes committed:
update edit text xpath to be read from any level

jostnes commented:
@fluiddot yup that looks to be it! I added the change in this commit https://github.com/WordPress/gutenberg/pull/40670/commits/92ffc5eeff491355eb8f8677b8e6d3de3853c1ab and looks like Android tests are green again 🎉  

There are some failing tests on iOS which I had just restarted, will wait for those to complete to see if those are flakiness or consistent failures.

fluiddot commented:
> @fluiddot yup that looks to be it! I added the change in this commit [92ffc5e](https://github.com/WordPress/gutenberg/commit/92ffc5eeff491355eb8f8677b8e6d3de3853c1ab) and looks like Android tests are green again 🎉

Awesome, happy to know that updating the locator addressed the issue 🎊. 

> There are some failing tests on iOS which I had just restarted, will wait for those to complete to see if those are flakiness or consistent failures.

@jostnes I did the last check and looks like all tests are 🟢 , so feel free to merge the PRs.
